File: C++/PathSumII.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<unordered_set>
#include<string>
#include<string.h>
#include<algorithm>
#include<math.h>
#include<map>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<vector<int>> ret;
	bool isLeaf(TreeNode* node)
	{
		if(node->left || node->right)
			return false;
		return true;
	}
	void dfs(TreeNode* root, int sum, vector<int> path)
	{
		if(!root)
			return;
		path.push_back(root->val);
		sum -= root->val;
		if(isLeaf(root))
		{
			if(sum == 0)
				ret.push_back(vector<int>(path));
		}
		else
		{
			dfs(root->left, sum, path);
			dfs(root->right, sum, path);
		}
		path.pop_back();
	}
	vector<vector<int> > pathSum(TreeNode *root, int sum) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		ret.clear();
		vector<int> path;
		dfs(root, sum, path);
		return ret;
	}
};
int main()
{
	TreeNode root(1);
	int sum = 1;
	Solution sol;
	sol.pathSum(&root, sum);
	getchar();
}

File: JAVA/PathSumII.java
import java.util.ArrayList;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}

class Solution {
	ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
	public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ret.clear();
		ArrayList<Integer> path = new ArrayList<Integer>();
		dfs(root, sum, path);
		return ret;
	}
	private void dfs(TreeNode root, int sum, ArrayList<Integer> path) {
		if(root == null)
			return;
		path.add(root.val);
		if(isLeaf(root))
		{
			if(sum == root.val)
			{
				ret.add((ArrayList<Integer>) path.clone());
			}
		}
		else
		{
			dfs(root.left, sum-root.val, path);
			dfs(root.right, sum-root.val, path);
		}
		path.remove(path.size()-1);
		
	}
	private boolean isLeaf(TreeNode root) {
		if(root == null)
			return false;
		if(root.left == null && root.right == null)
			return true;
		return false;
	}
}

public class PathSumII {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
